spec/system/admin/bulk_url_upload_spec.rb

```diff
@@ -1,27 +1,68 @@
 # frozen_string_literal: true
 
-describe 'Bulk URL upload' do
-  include ActiveJob::TestHelper
+shared_examples 'a successful bulk upload' do
+  it 'sends us back to the bulk upload page' do
+    do_bulk_upload
+    expect(page).to have_text('Bulk Search.gov URL Upload')
+  end
+
+  it 'shows a confirmation message' do
+    do_bulk_upload
+    expect(page).to have_text(
+      <<~CONFIRMATION_MESSAGE
+        Successfully uploaded #{upload_filename} for processing.
+        The results will be emailed to you.
+      CONFIRMATION_MESSAGE
+    )
+  end
 
-  subject(:bulk_upload) do
-    perform_enqueued_jobs do
-      visit url
-      attach_file('bulk_upload_urls', url_file)
-      click_button('Upload')
+  it 'creates the URLs' do
+    do_bulk_upload
+
+    urls.each do |url|
+      expect(SearchgovUrl.find_by(url: url)).not_to be_blank
     end
   end
 
+  it 're-indexes the domains for the URLs' do
+    do_bulk_upload
+    expect(@reindexed_domains).to eq(searchgov_domains)
+  end
+end
+
+shared_examples 'a failed bulk upload with error' do |error_message|
+  it 'sends us back to the bulk upload page' do
+    do_bulk_upload
+    expect(page).to have_text('Bulk Search.gov URL Upload')
+  end
+
+  it 'shows an error message' do
+    do_bulk_upload
+    expect(page).to have_text(error_message)
+  end
+end
+
+def do_bulk_upload
+  perform_enqueued_jobs do
+    visit url
+    attach_file('bulk_upload_urls', upload_file)
+    click_button('Upload')
+  end
+end
+
+describe 'Bulk URL upload' do
+  include ActiveJob::TestHelper
+
   let(:url) { '/admin/bulk_url_upload' }
-  let(:url_filedir) { 'txt' }
-  let(:url_filename) { 'good_url_file.txt' }
-  let(:url_file) { file_fixture("#{url_filedir}/#{url_filename}") }
-  let(:urls) { File.open(url_file).readlines.map(&:strip) }
+  let(:upload_file) { file_fixture("txt/#{upload_filename}") }
+  let(:urls) { File.open(upload_file, 'r:bom|utf-8').readlines.map(&:strip).map { |url| URI.escape(url) } }
```

Comment:
This line is a little too close to how the sausage is made, and it obscures what the spec is actually testing. (Note: we actually don't use `URI.escape` to encode the URLs; we use `Addressable::URI.normalized_encode`.) For the spec for this story, we can be be explicit about the resulting URLs in the database:
```ruby
  context 'when the URLs contain non-ASCII characters' do
    include_context 'log in super admin'

    let(:upload_filename) { 'utf8_urls.txt' }
    
    it 'encodes the URLs' do
      do_bulk_upload

      expect(SearchgovUrl.pluck(:url)).to include(
        "https://www.agency.gov/sites/default/files/advisory/2018-04-19/Asesor%E2%88%9A%E2%89%A0a%20en%20Venezuela%20sobre%20corrupci%E2%88%9A%E2%89%A5n%20(FIN-2017-A006).pdf",
        "https://www.agency.gov/web/oge.nsf/Legal%20Docs%20RSS/2C6DE2BBA325C32A852585BB005DA324/$FILE/LA-20-03%20Conflict%20of%20Interest%20Analysis%20for%20Stocks%20under%2018%20U.S.C.%20%C2%A7%20208.pdf?open"
      )
    end
  end
```
Note: we can also tighten up this spec by using simplified versions of the problematic URLs, i.e.
`https://www.agency.gov/foo § 208.pdf?open`

Reply:
Changed.

Reply:
Changed.

Reply:
Changed.